### PR TITLE
Allow Mapping and Identity to be fittable models

### LINKED ISSUE #6012: Make `Mapping` and `Identity` fittable"models
Currently `Mapping` and `Identity` are not "fittable" . This makes compound models whic include them not "fittable".  @jehturner

### PR DESCRIPTION
**Pull request by pllim**
Fix #6012 

c/c @nden @jehturner

### CODE REVIEW

**Comment by jehturner** [comment]
Dammit, @pllim, you're fast! I'll just try this out. Thanks a lot.

**Comment by pllim** [comment]
Ops, @jehturner , were you also working on the fix? Sorry, I didn't know!

Not sure why CircleCI failed, I couldn't see the log (only a spinning circle that won't stop spinning).

**Comment by jehturner** [comment]
No, not yet. I was just thinking about doing it :-). Are you in Leiden?

**Comment by pllim** [comment]
> Are you in Leiden?

No. I am just procrastinating from some IRAF related stuff. :sweat_smile:

**Comment by jehturner** [comment]
```
=================================== FAILURES ===================================
____________________________ test_fittable_compound ____________________________

    def test_fittable_compound():
        m = Identity(1) | Mapping((0, )) | Gaussian1D(1, 5, 4)
        x = np.arange(10)
        y_real = m(x)
        dy = 0.005
        y_noisy = y_real + np.random.normal(0., dy, x.shape)
        pfit = LevMarLSQFitter()
>       new_model = pfit(m, x, y_noisy)

astropy/modeling/tests/test_mappings.py:77: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <astropy.modeling.fitting.LevMarLSQFitter object at 0xf06ad10c>
model = <CompoundModel240(amplitude_2=1.0, mean_2=5.0, stddev_2=4.0)>
x = array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
y = array([ 0.45072537,  0.61055694,  0.76058391,  0.88834404,  0.96522319,
        1.00361239,  0.96949817,  0.8765893 ,  0.75145885,  0.60569953])
z = None, weights = None, maxiter = 100, acc = 1e-07
epsilon = 1.4901161193847656e-08, estimate_jacobian = False

    def __call__(self, model, x, y, z=None, weights=None,
                 maxiter=DEFAULT_MAXITER, acc=DEFAULT_ACC,
                 epsilon=DEFAULT_EPS, estimate_jacobian=False):
        """
            Fit data to this model.
    
            Parameters
            ----------
            model : `~astropy.modeling.FittableModel`
                model to fit to x, y, z
            x : array
               input coordinates
            y : array
               input coordinates
            z : array (optional)
               input coordinates
            weights : array (optional)
               weights
            maxiter : int
                maximum number of iterations
            acc : float
                Relative error desired in the approximate solution
            epsilon : float
                A suitable step length for the forward-difference
                approximation of the Jacobian (if model.fjac=None). If
                epsfcn is less than the machine precision, it is
                assumed that the relative errors in the functions are
                of the order of the machine precision.
            estimate_jacobian : bool
                If False (default) and if the model has a fit_deriv method,
                it will be used. Otherwise the Jacobian will be estimated.
                If True, the Jacobian will be estimated in any case.
    
            Returns
            -------
            model_copy : `~astropy.modeling.FittableModel`
                a copy of the input model with parameters set by the fitter
            """
    
>       from scipy import optimize
E       ImportError: No module named scipy

astropy/modeling/fitting.py:546: ImportError
= 1 failed, 11124 passed, 565 skipped, 49 xfailed, 1 xpassed in 225.15 seconds =

./.run_docker_tests.sh returned exit code 1
```

**Comment by pllim** [comment]
Thanks for the log. Looks like I'll have to add a marker for scipy.

**Comment by pllim** [comment]
Tests passed.

**Comment by jehturner** [comment]
When I try fitting a bivariate compound model (`Mapping((0, 1, 1)) | Chebyshev2D(2, 2) & Identity(1)`) with `LinearLSQFitter`, I get the following error. Is this not incorrect? The terms are still polynomial.
```
astropy.modeling.fitting.ModelLinearityError: Model is not linear in parameters, linear fit methods should not be used.
```
When I instead try using `LevMarLSQFitter()`, the code then runs, but all the fitted model coefficients (and thus values produced by the model) end up being zero. Is this because `LevMarLSQFitter()` expects to be given starting guesses for the coefficients, as with the `Gaussian` in your new test? I am fairly ignorant about non-linear fitting and if this is the intention, the docs don't seem to spell it out, but I gather it may be necessary in general. However, even if I try setting the Chebyshev coefficients before constructing and fitting the compound model from it, they come out completely unchanged by the "fit", whether or not the starting values were exactly correct.

Thanks!

**Comment by jehturner** [comment]
```
from astropy.modeling.models import *
from astropy.modeling.fitting import *
import numpy as np

# Using order (1, 1) here doesn't change anything:
simple = Chebyshev2D(2, 2)
compound = Mapping((0, 1, 1)) | Chebyshev2D(2, 2) & Identity(1)

#fitter = LinearLSQFitter()  # doesn't work
fitter = LevMarLSQFitter()

y, x = np.mgrid[0:3, 0:3]
z = x + y

simple_fit = Mapping((0, 1, 1)) | fitter(simple, x, y, z) & Identity(1)
simple_z = simple_fit(x, y)[0]
print simple_z

compound_fit = fitter(compound, x, y, z)
compound_z = compound_fit(x, y)[0]
print compound_z
```

**Comment by jehturner** [comment]
Sorry, I posted a (deleted) example that didn't make any sense, because I was trying to fit the bivariate function to a single array of z values. I'm not sure why the fitter doesn't complain about that, but you can probably ignore what I wrote about getting all zeroes above, assuming that's the reason. Am I right in thinking that this can't be done, because the fitters only accept a single z (and/or because it's inherently a crazy thing to do).

**Comment by jehturner** [comment]
Sorry, I posted a (deleted) example that didn't make any sense, because I was trying to fit the bivariate function to a single array of z values. I'm not sure why the fitter doesn't complain about that, but you can probably ignore what I wrote about getting all zeroes above, assuming that's the reason. Am I right in thinking that this can't be done, because the fitters only accept a single z (and/or because it's inherently a crazy thing to do)?

**Comment by jehturner** [comment]
When people fit astrometric distortions with `modeling`, do they just model each output co-ordinate separately and then tack them together with `mapping`? I'd have thought that could produce some slight inconsistencies between the axes. Or are people not actually doing astrometric fits? I assume models like SIP are there for that purpose, but maybe it's still an incomplete work-in-progress if one can't fit 2 outputs (as far as I can see)?

It seems like I'm not the only person asking roughly this question...
http://stackoverflow.com/questions/43548183/fitting-wcs-coordinate-solution-using-identified-stars-with-astropy
https://mail.python.org/pipermail/astropy/2016-November/004067.html

Sorry, this is getting a bit off topic for this PR, but I'm just trying to understand the scope of what is expected to work after Pey-Lian's fix and how to test it properly.

**Comment by bsipocz** [comment]
@pllim - If this is a bug, then you can still milestone it as 1.3.3. I plan to make on final release in the next couple of weeks.

**Comment by nden** [comment]
James, 

Fitting a transform and creating a WCS from matching pairs of (x,y) and (ra,dec) is a long standing request for GWCS. I understand it's going to be very useful, just haven't had any time to work on it. If anyone else has time, that'd be great. I see it as a replacement of geotran in iraf. 
I can't answer the question about calibration.

I am also not very familiar with the compound model fitting code. But I think in general a compound model should be fitted with a non-linear fitter because at least currently the LinearFitter cannot construct the Vandermonde matrix for compound models.

However I can confirm that using LevMarLSQFitter did not work with your example. It's something to look into and there might be an issue for this.

**Comment by jehturner** [comment]
Thanks! I can't spot an existing issue for fitting bivariate functions with 2 outputs (or astrometry specifically) -- shall I open one? Thanks for the info. about LinearFitter; I'll add that to the docs. I'm not sure which calibration question you're referring to.

**Comment by nden** [comment]
There's an astrometry issue filed at spacetelescope/gwcs#79. However an issue can be filed here for a fitter that works with models with more than one output, e.g. affine transforms. I misread your calibration question, I see now it was not a general question but specific to modeling.

**Comment by pllim** [comment]
So, is there anything else I need to do in this PR as far as enabling fitting in mapping?

**Comment by jehturner** [comment]
Your fix did get rid of the `ValueError: Model does not appear to be fittable`, thanks, but I was having trouble determining whether that actually fixed the compound model fitting, without a working 2D fitter (apparently it did work for your test case though). I was just going to try something slightly more contrived that doesn't depend on a 2D fitter, to establish whether fitting with a non-trivial mapping is inherently broken (if it is, do you want to treat that as a separate issue?).

**Comment by pllim** [comment]
> if it is, do you want to treat that as a separate issue?

Yes, if you still want these models to be fittable despite the problems you mentioned. This is because anything more complex will probably require @nden to fix anyway.

But if not getting your use-case to work is a deal-breaker from making these fittable, then no.

**Comment by jehturner** [comment]
I think that (as @nden seemed to have concluded), fitting certain compound cases with a Mapping still isn't working:
```
from astropy.modeling.models import *
from astropy.modeling.fitting import *
import numpy as np

# Using order (1, 1) here doesn't change anything:
simple = Chebyshev2D(2, 2)
compound = Mapping((0, 1, 1)) | Chebyshev2D(2, 2) & Identity(1)

fitter = LevMarLSQFitter()

y, x = np.mgrid[0:3, 0:3]
z = x + y

simple_fit = Mapping((0, 1, 1)) | fitter(simple, x, y, z) & Identity(1)
simple_z = simple_fit(x, y)[0]
print simple_z

compound_fit = fitter(compound | Mapping((0,), n_inputs=2), x, y, z) 
compound_z = compound_fit(x, y)
print compound_z
```
But if your fix doesn't break anything, it's a good start :smile:. Otherwise we'll have to do it later anyway before these cases can be fixed. I was just checking what happens when I try to fit `Identity` by itself, to be sure I hadn't persuaded you to drop some useful safeguard, but that case seems to do something sensible, returning the usual `Identity` mapping but with a warning that `The fit may be unsuccessful`.

Thanks!

James.

**Comment by jehturner** [comment]
I've added a version of the example above (with your changes; oops, bad habits) to the parent issue, if you want to merge the PR.

File: astropy/modeling/tests/test_mappings.py [hunk 1]
```diff
@@ -57,8 +66,21 @@ def test_identity():
...
-                                  [ 1.,  1.,  1.]])))
+                    (np.array([[-1.73205081, -1.73205081, -1.73205081],
+                               [-1.73205081, -1.73205081, -1.73205081]]),
+                     np.array([[1., 1., 1.],
+                               [1., 1., 1.]])))
     assert_allclose(model.inverse(res_x, res_y), (x, y), atol=1.e-10)
+
+
+@pytest.mark.skipif('not HAS_SCIPY')
+def test_fittable_compound():
```
Comment (nden) [inline astropy/modeling/tests/test_mappings.py]: @pllim Could you add the issue or PR number in a docstring. I have the feeling this test won't make much sense to me in a few months.

**Comment by jehturner** [comment]
I believe I've spotted a little complication here... This PR makes `Mapping` inherit from `FittableModel` instead of `Model`, but `FittableModel` defines `linear=False` by default, so the fix changes `Mapping((0,)).linear` and `Identity(1).linear` from `True` to `False`, which I assume was unintentional? I think that's the real reason for the first example error in #6038.

**Comment by pllim** [comment]
>  I think that's the real reason for the first example error

@jehturner , good catch. I made them linear
```python
>>> m = Mapping((0,)) | Chebyshev1D(2)
>>> m.linear
True
```
Your first example now fails with the same error as the second example.

File: astropy/modeling/tests/test_mappings.py (hunk 1 above)
Comment (pllim) [inline astropy/modeling/tests/test_mappings.py]: Sure. Done.

**Comment by pllim** [comment]
I think just making them fittable is non-controversial enough. Whether the fitting actually make sense should be a separate discussion. Merging this.

@bsipocz I think this is okay as v2.0 since there are still some follow-up discussions to be had.

**Pull request merged**